fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java

```diff
@@ -3087,8 +3091,124 @@ private void setOperationContextProperties(FHIROperationContext operationContext
+        int result = 0;
+        // Since the try logic is slightly different in the code paths, we want to dispatch to separate methods to simplify the logic.
+        if (resourceLogicalId != null) {
+            result = doReindexSingle(operationOutcomeResult, tstamp, resourceLogicalId);
+        } else {
+            result = doReindexList(operationOutcomeResult, tstamp, indexIds);
+        }
+        return result;
+    }
+
+    /**
+     * encapsulates the logic to process a list with graduated backoff through the full list of indexIds
+     *
+     * @param operationOutcomeResult
+     * @param tstamp
+     * @param indexIds
+     * @return
+     * @throws Exception
+     */
+    public int doReindexList(OperationOutcome.Builder operationOutcomeResult, Instant tstamp, List<Long> indexIds) throws Exception {
+        /*
+         * How the backoff works...
+         * indexIds[1,2,3,4,5,6,7,8,9,10]
+         *
+         * Pass 1 left=0, right=10, max=10
+         * Result: Deadlock
+         *
+         * Pass2 left=0, right=1, max=10
+         * Result: Pass
+         *
+         * Pass3 left=1, right=10, max=10
+         * Result: Deadlock
+         *
+         * Pass 4: left=1, right=2, max=10
+         * Result: Pass
+         *
+         * Pass 5: left=2, right=10, max=10
+         * Result: Pass
+         *
+         * Return total count back to caller.
+         */
+
+        // Maximum number to retry
+        final int TX_ATTEMPTS = 5;
+
+        int result = 0;
+        int max = indexIds.size();
+
+        int left = 0;
+        int right = max;
+
+        int window = 0;
+        while (left < max) {
+            window++;
+            if (log.isLoggable(Level.FINE)) {
+                log.fine("$reindex window[" + window + "] -> left=[" + left + "] right=[" + right + "] max=[" + max + "]");
+            }
+
+            boolean backoff = false;
+            List<Long> subListIndexIds = indexIds.subList(left, right);
+
+            int attempt = 1;
+            while (attempt <= TX_ATTEMPTS && !backoff) {
+                FHIRTransactionHelper txn = new FHIRTransactionHelper(getTransaction());
+                txn.begin();
+                try {
+                    FHIRPersistenceContext persistenceContext = null;
+                    result += persistence.reindex(persistenceContext, operationOutcomeResult, tstamp, subListIndexIds, null);
+                    attempt = TX_ATTEMPTS + 1; // end the retry loop
+                } catch (FHIRPersistenceDataAccessException x) {
+                    // At this point, the transaction is marked for rollback
+                    if (x.isTransactionRetryable()) {
+                        if (x.getCause() instanceof LockException) {
+                            backoff = true;
+                            long wait = 0;
+                            log.fine(() -> "'DeadLock' executing random retry and backing off on processing");
+                            wait = RANDOM.nextInt(5000);
```

Comment:
Don't see where you actually do the wait.

Reply:
thanks, fixed in latest local code, just testing now.

Reply:
code is pushed